concent_api/core/models.py

```diff
@@ -73,3 +75,216 @@ class MessageAuth(Model):
     requestor_public_key = Base64Field(max_length=64)
+
+
+class ClientManager(Manager):
+
+    def get_or_create_full_check(self, public_key):
+        """
+        Returns Model instance.
+        Does the same as get_or_create method, but also performs full_clean() on newly created instance.
+        """
+        try:
+            instance = self.get(public_key = public_key)
+        except self.model.DoesNotExist:
+            instance = self.model(
+                public_key = public_key
+            )
+            instance.full_clean()
+            instance.save()
+        return instance
+
+
+class Client(models.Model):
+    """
+    Represents Concent client, identified by public key.
+    """
+
+    objects = ClientManager()
+
+    public_key = Base64Field(max_length = 64, unique = True)
+
+
+class Subtask(models.Model):
+    """
+    Represents subtasks states.
+    """
+
+    class SubtaskState(ChoiceEnum):
+        ForcingReport           = 'forcing_report'
+        Reported                = 'reported'
+        ForcingResultTransfer   = 'forcing_result_transfer'
+        ResultUploaded          = 'result_uploaded'
+        ForcingAcceptance       = 'forcing_acceptance'
+        Rejected                = 'rejected'
+        AdditionalVerification  = 'additional_verification'
+        Accepted                = 'accepted'
+        Failed                  = 'failed'
+
+    class Meta:
+        unique_together = (
+            ('requestor', 'task_id'),
+            ('requestor', 'subtask_id'),
+        )
+
+    task_id                     = models.CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH)
+    subtask_id                  = models.CharField(max_length = MESSAGE_TASK_ID_MAX_LENGTH)
+
+    # Relation to concent client in requestor context.
+    requestor                   = models.ForeignKey(Client, related_name = 'subtasks_as_requestor')
+
+    # Relation to concent client in provider context.
+    provider                    = models.ForeignKey(Client, related_name = 'subtasks_as_provider')
+
+    state                       = models.CharField(max_length = 32, choices = SubtaskState.choices())
+
+    # If in an active state, it's the time at which Concent automatically transitions to a passive state if it does
+    # not get the information it expects from the client (a timeout). Must be NULL in a passive state.
+    next_deadline               = models.DateTimeField(null = True)
+
+    # Related messages
+    task_to_compute             = models.ForeignKey(StoredMessage, null = True)
+    report_computed_task        = models.ForeignKey(StoredMessage, null = True)
+    ack_report_computed_task    = models.ForeignKey(StoredMessage, null = True)
+    reject_report_computed_task = models.ForeignKey(StoredMessage, null = True)
+    subtask_results_accepted    = models.ForeignKey(StoredMessage, null = True)
+    subtask_results_rejected    = models.ForeignKey(StoredMessage, null = True)
+
+
+class ReceiveQueue(models.Model):
```

Comment:
A model represents a single row of a table. Naming it `ReceiveQueue` would suggest that every object is a separate queue which is not true.